switch from time to systemtime

esheppa opened the pull request:
This reduces a dependency, but also makes `ulid` compatible with all datetime libraries as the internal types generally implement `From<SystemTime>`.

This should also solve #52 and potentially #50

dylanhart reviewed (changes requested):
Hi Eric! Thank you for you contribution. Currently the PR is not compiling:

```
$ cargo test
   Compiling ulid v0.6.0
error[E0433]: failed to resolve: use of undeclared type `Duration`
   --> src\generator.rs:182:46
    |
182 |         let ulid3 = Ulid::from_datetime(dt + Duration::from_millis(1));
    |                                              ^^^^^^^^ not found in this sc
ope
    |
help: consider importing one of these items
    |
174 |     use core::time::Duration;
    |
174 |     use std::time::Duration;
    |

For more information about this error, try `rustc --explain E0433`.
error: could not compile `ulid` due to previous error
warning: build failed, waiting for other jobs to finish...
```

Can you fix the issues? The following should all pass:

```
cargo test
cargo test --all-features
cargo test --no-default-features --features=std
cargo test --no-default-features
```

Thanks!

esheppa commented:
Thanks for the review @dylanhart - I've fixed the compile error and verified with the `cargo test` commands you listed. I also fixed the cli which I'd broken due to `SystemTime` not implementing `Display`, so I've added that to the CI checks as well

dylanhart commented on src/generator.rs:
```diff
@@ -120,16 +117,21 @@ impl Generator {
-        if (datetime.unix_timestamp_nanos() / 1_000_000) <= last_ms {
+        if datetime
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
```
Let's not panic. Instead we can saturate at zero with: `.unwrap_or(Duration::ZERO)`.

esheppa replied:
Good idea - I've also documented the behavior for `SystemTime`'s earlier than the unix epoch and added some tests to verify the behavior

dylanhart commented on src/lib.rs:
```diff
@@ -145,13 +145,13 @@ impl Ulid {
     /// let ulid = Ulid::from_datetime(dt);
     ///
-    /// assert_eq!(ulid.timestamp_ms(), (dt.unix_timestamp_nanos() / 1_000_000) as u64);
+    /// assert_eq!(u128::from(ulid.timestamp_ms()), dt.duration_since(SystemTime::UNIX_EPOCH).unwrap().as_millis());
```
`.unwrap_or(Duration::ZERO)`

dylanhart commented on src/time.rs:
```diff
@@ -35,31 +34,34 @@ impl Ulid {
-        let timestamp = datetime.unix_timestamp_nanos() / 1_000_000;
+        let timestamp = datetime
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
```
`.unwrap_or(Duration::ZERO)`

dylanhart reviewed (changes requested):
Hi Eric! Thanks for the update. Can you remove the panicking `.unwrap()` calls? Thanks!

esheppa commented:
Thanks @dylanhart  - I've fixed those `unwrap` calls - let me know if you are ok with the additional wording in the docs

dylanhart commented on Cargo.toml:
```diff
@@ -14,13 +14,13 @@ repository = "https://github.com/dylanhart/ulid-rs"
 default = ["std"]
-std = ["rand", "time"]
+std = ["rand"]
+time = []
```
Why was `time` added as a feature here? I'm not seeing anything that uses it. I think we can remove this and the reference to `time` (the library) from README.md. :)

esheppa replied:
I'd thought this is required as users may have selected: `features = ["time"]`, however on second thoughts, this is a semver incompatible change anyway so it's fine to have breakage here as well

dylanhart reviewed (changes requested):
This looks great. The code and rustdoc look good to me. Just one comment on cargo features.

esheppa commented:
I've cleaned up the Cargo.toml and README.md. I've left the benchmarks as is as I assume you'd like to run them on your computer for consistency

dylanhart commented:
Looks good! Thanks! Will make a release for this later.

dylanhart commented:
Available in v1.0.0
